server/src/main/java/org/opensearch/gateway/ShardsBatchGatewayAllocator.java

```diff
@@ -516,8 +516,33 @@ protected AsyncShardFetch.FetchResult<TransportNodesListShardStoreMetadataBatch.
 
         @Override
         protected boolean hasInitiatedFetching(ShardRouting shard) {
+            /**
+             * This function is to check if asyncFetch has happened before for this shard batch, or is ongoing.
+             * It should return false if there has never been a fetch for this batch.
+             * This function is currently only used in the case of replica shards when all deciders returned NO/THROTTLE, and explain mode is ON.
+             * Allocation explain and manual reroute APIs try to append shard store information (matching bytes) to the allocation decision.
+             * However, these APIs do not want to trigger a new asyncFetch for these ineligible shards, unless the data from nodes is already there.
+             * This function is used to see if a fetch has happened to decide if it is possible to append shard store info without a new async fetch.
+             *
+             * In order to check if a fetch has ever happened, we check 2 things:
+             * 1. If the shard batch cache is empty, we know that fetch has never happened so we return false.
+             * 2. If we see that the list of nodes to fetch from is empty, we know that all nodes have data or are ongoing a fetch. So we return true.
+             * 3. Otherwise we return false.
+             *
+             * see {@link AsyncShardFetchCache#findNodesToFetch()}
+             */
             String batchId = getBatchId(shard, shard.primary());
-            return batchId != null;
+            logger.trace("Checking if fetching done for batch id {}", batchId);
+            ShardsBatch shardsBatch = shard.primary() ? batchIdToStartedShardBatch.get(batchId) : batchIdToStoreShardBatch.get(batchId);
+
+            // if fetchData has never been called, the per node cache will be empty and have no nodes
+            // this is because cache.fillShardCacheWithDataNodes(nodes) initialises this map and is called in AsyncShardFetch.fetchData
+            if (shardsBatch.getAsyncFetcher().hasEmptyCache()) {
+                logger.trace("Batch cache is empty for batch {} ", batchId);
+                return false;
+            }
+            // this check below is to make sure we already have all the data and that we wouldn't create a new async fetchData call
```

Comment:
if we have all the data from previous fetch, then also it should be true

Reply:
We are checking that in the below line 

`return shardsBatch.getAsyncFetcher().getCache().findNodesToFetch().isEmpty();`

This would return true if we have all the data from previous fetch, or if fetch has just triggered

Reply:
Could there be a case where some of the nodes are not yet initialized in the Cache. I see it used in tandem in other place where this method is used

```
        cache.fillShardCacheWithDataNodes(nodes);
        List<String> nodeIds = cache.findNodesToFetch();
```

Reply:
is if (shardsBatch.getAsyncFetcher().hasEmptyCache()) { enough. do we need the check on line no. 545?

Reply:
@shwetathareja I don't think it's enough. There might be a case where the cache has entries for all nodes except a few (say node joined or node left).

 We need to verify that every node has had a fetchData before or in progress - so that In this case, the cache would not be empty, but it would still have entries missing for a few nodes.

Whenever this method is called, we need to make sure we don't trigger a new asyncFetch at all. We want to only use the existing data available to populate the shard store information along with allocation decision NO/THROTTLE.
What we actually want to check is that if we called fetchData now, would it trigger a new asyncFetch.

So in a case like this, we need to also see if **every node** has an entry or has a fetch ongoing - we shouldn't just rely on the cache being empty or not, because it doesn't give the full picture of all the nodes.

Reply:
@rajiv-kv yeah in this function we are essentially relying on the fact that this code below has run before.
I've mentioned details about this in the PR description.

`hasInitiatedFetching` is more of a check to see if every node has completed or initiated fetching

```        
cache.fillShardCacheWithDataNodes(nodes);
List<String> nodeIds = cache.findNodesToFetch();
```

If some nodes are not initialised in the cache, it means that `cache.findNodesToFetch()` would return a non empty list. In this case, it means our cache does not have all the info available, so if we were to call `fetchData`, it would trigger a new `asyncFetch`. 

We don't want to trigger a new asyncFetch in allocation explain API in this case. So `hasInitiatedFetching` function should return false so that we exit early with decision NO.

Reply:
I think it is fair to assume that node-join and node-left will keep the node-list updated in cache.
However there is also a updated node-list in the input(`RoutingAllocation#nodes`) that can be used to compare against the cache to avoid staleness.

Reply:
Makes sense. I think in most cases, it should be fair to assume that `if(shardsBatch.getAsyncFetcher().hasEmptyCache())` check is sufficient since the nodes list will be updated.
I had just added this extra check to be 100% sure we don't trigger any new `asyncFetch`

On the `RoutingAllocation#nodes` suggestion, are you suggesting that in addition to this, we also check to see if the `RoutingAllocation#nodes` object has any new nodes compared to the list of nodes in the cache?

Reply:
Yes i was thinking of checking instead of `findNodesToFetch`. Something like

```allocation#nodes.foreach{node-> cache.hasFetchData(node)}```

Reply:
Got it.

I checked the node join and node left flows again.
They are triggered with `explain=false` so `hasInitiatedFetching()` would not get triggered.
They are also URGENT priority tasks, so they would keep the nodes up to date.

I think we don't need to check for stale nodes as of now and the current implementation is fine - we just assume that the nodes list is up to date in the cache and check that every node has an entry, or is actively fetching.

If we do see an edge case for this with node-joins/node-lefts , maybe we can take it up as a follow up issue.